File: frontend/src/api/types/grant.ts
// src/api/types/grant.ts
//
// Types aligned with the /grants router (grants_router.py).

// Body you POST to /grants
export interface GrantCreate {
  resource_id: string;
  grantee_type?: string;        // "user" | "invite"
  grantee_id?: string;          // user_id for direct grant; omit for invite
  // CRUDEASIO permission flags
  can_create?: boolean;
  can_read?: boolean;
  can_update?: boolean;
  can_delete?: boolean;
  can_evict?: boolean;
  can_invoke?: boolean;
  can_add?: boolean;
  can_share?: boolean;
  can_admin?: boolean;
  // Invite targeting (optional)
  target_entity?: string;
  target_entity_type?: string;
  max_claims?: number;
  requires_identity?: boolean;
  name?: string;
  notes?: string;
  expires_at?: string;
  state?: string;
}

// Body you PATCH to /grants/{grant_id}
export interface GrantUpdate {
  max_claims?: number;
  state?: string;
  name?: string;
  notes?: string;
  expires_at?: string;
  can_create?: boolean;
  can_read?: boolean;
  can_update?: boolean;
  can_delete?: boolean;
  can_evict?: boolean;
  can_invoke?: boolean;
  can_add?: boolean;
  can_share?: boolean;
  can_admin?: boolean;
}

// What grant endpoints return
export interface GrantResponse {
  id: string;
  resource_id: string;
  grantee_type: string;
  grantee_id: string;
  granted_by: string;
  can_create: boolean;
  can_read: boolean;
  can_update: boolean;
  can_delete: boolean;
  can_evict: boolean;
  can_invoke: boolean;
  can_add: boolean;
  can_share: boolean;
  can_admin: boolean;
  requires_identity: boolean;
  read_requires_identity?: boolean | null;
  write_requires_identity?: boolean | null;
  invoke_requires_identity?: boolean | null;
  target_entity?: string | null;
  target_entity_type?: string | null;
  max_claims?: number | null;
  claims_count: number;
  state: string;
  name?: string | null;
  notes?: string | null;
  granted_at: string;
  expires_at?: string | null;
  revoked_at?: string | null;
  revoked_by?: string | null;
  accepted_at?: string | null;
  accepted_by?: string | null;
  created_time: string;
  modified_time: string;
  // Only present on invite creation response
  claim_token?: string;
}

export type Grant = GrantResponse;

File: src/origin/models/grant.py
"""Grant — replaces the Arango `grants` collection.

The hot path: `(grantee_id, resource_id, state)` is the primary access lookup.
`resource_id` references artifacts in Mantle's Arango DB and is intentionally
NOT a foreign key — we trade referential integrity for cross-DB independence.

CRUDEASIO permission model: Create, Read, Update, Delete, Evict, Add, Share,
Invoke, Admin.
"""

import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, Column, DateTime, Index, Integer, String, Text, Uuid, text

from origin.db.base import Base


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class Grant(Base):
    __tablename__ = "grants"

    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    resource_id = Column(Uuid, nullable=False, index=True)
    grantee_type = Column(String(16), nullable=False)
    grantee_id = Column(String(256), nullable=False)
    granted_by = Column(Uuid, nullable=False)
    effect = Column(String(8), nullable=False, default="allow")

    # CRUDEASIO flags
    can_create = Column(Boolean, nullable=False, default=False)
    can_read = Column(Boolean, nullable=False, default=True)
    can_update = Column(Boolean, nullable=False, default=False)
    can_delete = Column(Boolean, nullable=False, default=False)
    can_evict = Column(Boolean, nullable=False, default=False)
    can_invoke = Column(Boolean, nullable=False, default=False)
    can_add = Column(Boolean, nullable=False, default=False)
    can_share = Column(Boolean, nullable=False, default=False)
    can_admin = Column(Boolean, nullable=False, default=False)

    # Identity-required overrides (nullable: inherit from `requires_identity`)
    requires_identity = Column(Boolean, nullable=False, default=False)
    read_requires_identity = Column(Boolean, nullable=True)
    write_requires_identity = Column(Boolean, nullable=True)
    invoke_requires_identity = Column(Boolean, nullable=True)

    # Invite targeting (only used when grantee_type == 'invite')
    target_entity = Column(String(256), nullable=True)
    target_entity_type = Column(String(32), nullable=True)
    max_claims = Column(Integer, nullable=True)
    claims_count = Column(Integer, nullable=False, default=0)

    # Lifecycle
    state = Column(String(16), nullable=False, default="active", index=True)
    name = Column(String(256), nullable=True)
    notes = Column(Text, nullable=True)
    granted_at = Column(DateTime(timezone=True), nullable=True)
    expires_at = Column(DateTime(timezone=True), nullable=True, index=True)
    accepted_by = Column(Uuid, nullable=True)
    accepted_at = Column(DateTime(timezone=True), nullable=True)
    revoked_by = Column(Uuid, nullable=True)
    revoked_at = Column(DateTime(timezone=True), nullable=True)
    created_time = Column(DateTime(timezone=True), nullable=False, default=_utcnow)
    modified_time = Column(
        DateTime(timezone=True), nullable=False, default=_utcnow, onupdate=_utcnow
    )

    __table_args__ = (
        Index("ix_grants_grantee_resource_state", "grantee_id", "resource_id", "state"),
        Index("ix_grants_grantee_type_id_state", "grantee_type", "grantee_id", "state"),
        Index(
            "ix_grants_resource_active",
            "resource_id",
            "state",
            sqlite_where=text("state = 'active'"),
        ),
    )
